File: app/csharp/AccountingSystem.Tests/Repositories/AutoJournalRepositoryTest.cs
using AccountingSystem.Infrastructure.Entities;
using AccountingSystem.Infrastructure.Repositories;
using FluentAssertions;
using Npgsql;
using Testcontainers.PostgreSql;
using Xunit;

namespace AccountingSystem.Tests.Repositories;

/// <summary>
/// 自動仕訳リポジトリ - 統合テスト
/// </summary>
public class AutoJournalRepositoryTest : IAsyncLifetime
{
    private readonly PostgreSqlContainer _postgres;
    private TestDatabase? _testDb;
    private AutoJournalRepository? _repository;

    public AutoJournalRepositoryTest()
    {
        _postgres = new PostgreSqlBuilder()
            .WithImage("postgres:16-alpine")
            .WithDatabase("testdb")
            .WithUsername("testuser")
            .WithPassword("testpass")
            .Build();
    }

    public async Task InitializeAsync()
    {
        await _postgres.StartAsync();
        _testDb = new TestDatabase(_postgres);
        await _testDb.StartAsync();
        _repository = new AutoJournalRepository(_postgres.GetConnectionString());
    }

    public async Task DisposeAsync()
    {
        await _testDb!.StopAsync();
        await _postgres.DisposeAsync();
    }

    private async Task CleanupAsync()
    {
        await using var conn = new NpgsqlConnection(_postgres.GetConnectionString());
        await conn.OpenAsync();

        await using var cmd1 = new NpgsqlCommand(@"TRUNCATE TABLE ""自動仕訳実行ログ"" CASCADE", conn);
        await cmd1.ExecuteNonQueryAsync();

        await using var cmd2 = new NpgsqlCommand(@"TRUNCATE TABLE ""自動仕訳パターン明細"" CASCADE", conn);
        await cmd2.ExecuteNonQueryAsync();

        await using var cmd3 = new NpgsqlCommand(@"TRUNCATE TABLE ""自動仕訳パターン"" CASCADE", conn);
        await cmd3.ExecuteNonQueryAsync();

        await using var cmd4 = new NpgsqlCommand(@"TRUNCATE TABLE ""自動仕訳管理"" CASCADE", conn);
        await cmd4.ExecuteNonQueryAsync();
    }

    private async Task InsertTestAccountAsync(string accountCode, string accountName)
    {
        await using var conn = new NpgsqlConnection(_postgres.GetConnectionString());
        await conn.OpenAsync();

        await using var cmd = new NpgsqlCommand(@"
            INSERT INTO ""勘定科目マスタ"" (""勘定科目コード"", ""勘定科目名"", ""勘定科目種別"", ""合計科目"")
            VALUES (@code, @name, '資産', false)
            ON CONFLICT (""勘定科目コード"") DO NOTHING
        ", conn);
        cmd.Parameters.AddWithValue("code", accountCode);
        cmd.Parameters.AddWithValue("name", accountName);
        await cmd.ExecuteNonQueryAsync();
    }

    #region 自動仕訳管理テスト

    [Fact(DisplayName = "自動仕訳管理_登録と取得ができる")]
    public async Task TestManagement_InsertAndFind()
    {
        await CleanupAsync();

        var management = new AutoJournalManagement
        {
            SourceTableName = "売上データ",
            LastProcessedAt = new DateTime(2025, 1, 21, 10, 0, 0)
        };

        var id = await _repository!.InsertManagementAsync(management);
        id.Should().BeGreaterThan(0);

        var found = await _repository.FindManagementBySourceTableAsync("売上データ");
        found.Should().NotBeNull();
        found!.SourceTableName.Should().Be("売上データ");
        found.LastProcessedAt.Should().Be(new DateTime(2025, 1, 21, 10, 0, 0));
    }

    [Fact(DisplayName = "自動仕訳管理_最終処理日時を更新できる")]
    public async Task TestManagement_UpdateLastProcessedAt()
    {
        await CleanupAsync();

        var management = new AutoJournalManagement
        {
            SourceTableName = "給与データ",
            LastProcessedAt = new DateTime(2025, 1, 1, 0, 0, 0)
        };

        var id = await _repository!.InsertManagementAsync(management);

        var newProcessedAt = new DateTime(2025, 1, 21, 15, 30, 0);
        await _repository.UpdateLastProcessedAtAsync(id, newProcessedAt);

        var found = await _repository.FindManagementBySourceTableAsync("給与データ");
        found!.LastProcessedAt.Should().Be(newProcessedAt);
    }

    #endregion

    #region 自動仕訳パターンテスト

    [Fact(DisplayName = "自動仕訳パターン_明細を含めて登録・取得できる")]
    public async Task TestPattern_InsertAndFindWithDetails()
    {
        await CleanupAsync();
        await InsertTestAccountAsync("1300", "売掛金");
        await InsertTestAccountAsync("4100", "売上");
        await InsertTestAccountAsync("2120", "仮受消費税");

        var pattern = new AutoJournalPattern
        {
            PatternCode = "SALES_001",
            PatternName = "売上仕訳パターン",
            SourceTableName = "売上データ",
            Description = "売上データから仕訳を自動生成",
            IsActive = true,
            Details = new List<AutoJournalPatternDetail>
            {
                new AutoJournalPatternDetail
                {
                    AutoJournalPatternId = 0, // 登録時に設定される
                    LineNumber = 1,
                    DebitCreditFlag = "D",
                    AccountCode = "1300",
                    AmountExpression = "税込金額",
                    DescriptionTemplate = "売上: {顧客名}"
                },
                new AutoJournalPatternDetail
                {
                    AutoJournalPatternId = 0,
                    LineNumber = 2,
                    DebitCreditFlag = "C",
                    AccountCode = "4100",
                    AmountExpression = "税抜金額",
                    DescriptionTemplate = "売上: {顧客名}"
                },
                new AutoJournalPatternDetail
                {
                    AutoJournalPatternId = 0,
                    LineNumber = 3,
                    DebitCreditFlag = "C",
                    AccountCode = "2120",
                    AmountExpression = "消費税額",
                    DescriptionTemplate = "消費税: {顧客名}"
                }
            }
        };

        var patternId = await _repository!.InsertPatternAsync(pattern);
        patternId.Should().BeGreaterThan(0);

        var found = await _repository.FindPatternByIdAsync(patternId);
        found.Should().NotBeNull();
        found!.PatternCode.Should().Be("SALES_001");
        found.PatternName.Should().Be("売上仕訳パターン");
        found.IsActive.Should().BeTrue();
        found.Details.Should().HaveCount(3);
        found.Details[0].DebitCreditFlag.Should().Be("D");
        found.Details[1].DebitCreditFlag.Should().Be("C");
    }

    [Fact(DisplayName = "自動仕訳パターン_パターンコードで取得できる")]
    public async Task TestPattern_FindByCode()
    {
        await CleanupAsync();
        await InsertTestAccountAsync("1100", "現金");

        var pattern = new AutoJournalPattern
        {
            PatternCode = "CASH_001",
            PatternName = "現金仕訳パターン",
            SourceTableName = "現金データ",
            IsActive = true,
            Details = new List<AutoJournalPatternDetail>
            {
                new AutoJournalPatternDetail
                {
                    AutoJournalPatternId = 0,
                    LineNumber = 1,
                    DebitCreditFlag = "D",
                    AccountCode = "1100",
                    AmountExpression = "金額"
                }
            }
        };

        await _repository!.InsertPatternAsync(pattern);

        var found = await _repository.FindPatternByCodeAsync("CASH_001");
        found.Should().NotBeNull();
        found!.PatternName.Should().Be("現金仕訳パターン");
    }

    [Fact(DisplayName = "自動仕訳パターン_有効なパターン一覧を取得できる")]
    public async Task TestPattern_FindActivePatterns()
    {
        await CleanupAsync();
        await InsertTestAccountAsync("1100", "現金");

        // 有効なパターン
        await _repository!.InsertPatternAsync(new AutoJournalPattern
        {
            PatternCode = "ACTIVE_001",
            PatternName = "有効パターン1",
            SourceTableName = "テスト",
            IsActive = true,
            Details = new List<AutoJournalPatternDetail>
            {
                new AutoJournalPatternDetail
                {
                    AutoJournalPatternId = 0,
                    LineNumber = 1,
                    DebitCreditFlag = "D",
                    AccountCode = "1100",
                    AmountExpression = "金額"
                }
            }
        });

        await _repository.InsertPatternAsync(new AutoJournalPattern
        {
            PatternCode = "ACTIVE_002",
            PatternName = "有効パターン2",
            SourceTableName = "テスト",
            IsActive = true,
            Details = new List<AutoJournalPatternDetail>
            {
                new AutoJournalPatternDetail
                {
                    AutoJournalPatternId = 0,
                    LineNumber = 1,
                    DebitCreditFlag = "D",
                    AccountCode = "1100",
                    AmountExpression = "金額"
                }
            }
        });

        // 無効なパターン
        await _repository.InsertPatternAsync(new AutoJournalPattern
        {
            PatternCode = "INACTIVE_001",
            PatternName = "無効パターン",
            SourceTableName = "テスト",
            IsActive = false,
            Details = new List<AutoJournalPatternDetail>
            {
                new AutoJournalPatternDetail
                {
                    AutoJournalPatternId = 0,
                    LineNumber = 1,
                    DebitCreditFlag = "D",
                    AccountCode = "1100",
                    AmountExpression = "金額"
                }
            }
        });

        var activePatterns = (await _repository.FindActivePattersAsync()).ToList();
        activePatterns.Should().HaveCount(2);
        activePatterns.Should().OnlyContain(p => p.IsActive);
    }

    [Fact(DisplayName = "自動仕訳パターン_削除時に明細とログもカスケード削除される")]
    public async Task TestPattern_CascadeDelete()
    {
        await CleanupAsync();
        await InsertTestAccountAsync("1100", "現金");

        var pattern = new AutoJournalPattern
        {
            PatternCode = "DELETE_TEST",
            PatternName = "削除テスト",
            SourceTableName = "テスト",
            IsActive = true,
            Details = new List<AutoJournalPatternDetail>
            {
                new AutoJournalPatternDetail
                {
                    AutoJournalPatternId = 0,
                    LineNumber = 1,
                    DebitCreditFlag = "D",
                    AccountCode = "1100",
                    AmountExpression = "金額"
                }
            }
        };

        var patternId = await _repository!.InsertPatternAsync(pattern);

        // 実行ログを追加
        await _repository.InsertExecutionLogAsync(new AutoJournalExecutionLog
        {
            AutoJournalPatternId = patternId,
            ExecutedAt = DateTime.Now,
            ProcessedCount = 10,
            GeneratedCount = 10,
            Status = "SUCCESS"
        });

        // パターン削除
        await _repository.DeletePatternAsync(patternId);

        // 削除確認
        var found = await _repository.FindPatternByIdAsync(patternId);
        found.Should().BeNull();

        var logs = await _repository.FindExecutionLogsByPatternIdAsync(patternId);
        logs.Should().BeEmpty();
    }

    #endregion

    #region 自動仕訳実行ログテスト

    [Fact(DisplayName = "自動仕訳実行ログ_登録と取得ができる")]
    public async Task TestExecutionLog_InsertAndFind()
    {
        await CleanupAsync();
        await InsertTestAccountAsync("1100", "現金");

        var patternId = await _repository!.InsertPatternAsync(new AutoJournalPattern
        {
            PatternCode = "LOG_TEST",
            PatternName = "ログテスト",
            SourceTableName = "テスト",
            IsActive = true,
            Details = new List<AutoJournalPatternDetail>
            {
                new AutoJournalPatternDetail
                {
                    AutoJournalPatternId = 0,
                    LineNumber = 1,
                    DebitCreditFlag = "D",
                    AccountCode = "1100",
                    AmountExpression = "金額"
                }
            }
        });

        var log = new AutoJournalExecutionLog
        {
            AutoJournalPatternId = patternId,
            ExecutedAt = new DateTime(2025, 1, 21, 10, 0, 0),
            ProcessedCount = 100,
            GeneratedCount = 98,
            Status = "SUCCESS",
            Message = "正常終了"
        };

        var logId = await _repository.InsertExecutionLogAsync(log);
        logId.Should().BeGreaterThan(0);

        var logs = (await _repository.FindExecutionLogsByPatternIdAsync(patternId)).ToList();
        logs.Should().HaveCount(1);
        logs[0].ProcessedCount.Should().Be(100);
        logs[0].GeneratedCount.Should().Be(98);
        logs[0].Status.Should().Be("SUCCESS");
    }

    [Fact(DisplayName = "自動仕訳実行ログ_最新のログを取得できる")]
    public async Task TestExecutionLog_FindLatest()
    {
        await CleanupAsync();
        await InsertTestAccountAsync("1100", "現金");

        var patternId = await _repository!.InsertPatternAsync(new AutoJournalPattern
        {
            PatternCode = "LATEST_TEST",
            PatternName = "最新ログテスト",
            SourceTableName = "テスト",
            IsActive = true,
            Details = new List<AutoJournalPatternDetail>
            {
                new AutoJournalPatternDetail
                {
                    AutoJournalPatternId = 0,
                    LineNumber = 1,
                    DebitCreditFlag = "D",
                    AccountCode = "1100",
                    AmountExpression = "金額"
                }
            }
        });

        // 古いログ
        await _repository.InsertExecutionLogAsync(new AutoJournalExecutionLog
        {
            AutoJournalPatternId = patternId,
            ExecutedAt = new DateTime(2025, 1, 20, 10, 0, 0),
            ProcessedCount = 50,
            GeneratedCount = 50,
            Status = "SUCCESS"
        });

        // 新しいログ
        await _repository.InsertExecutionLogAsync(new AutoJournalExecutionLog
        {
            AutoJournalPatternId = patternId,
            ExecutedAt = new DateTime(2025, 1, 21, 15, 0, 0),
            ProcessedCount = 100,
            GeneratedCount = 100,
            Status = "SUCCESS",
            Message = "最新の処理"
        });

        var latest = await _repository.FindLatestExecutionLogAsync(patternId);
        latest.Should().NotBeNull();
        latest!.ProcessedCount.Should().Be(100);
        latest.Message.Should().Be("最新の処理");
    }

    #endregion
}

File: app/fsharp/AccountingSystem.Tests/Repositories/AutoJournalRepositoryTest.fs
module AccountingSystem.Tests.Repositories.AutoJournalRepositoryTest

open System
open Xunit
open FsUnit.Xunit
open Npgsql
open AccountingSystem.Domain.Models.AutoJournalPattern
open AccountingSystem.Domain.Types
open AccountingSystem.Application.Repositories
open AccountingSystem.Infrastructure.Repositories
open AccountingSystem.Tests.DatabaseTestBase

/// <summary>
/// AutoJournalRepository の統合テスト
/// Testcontainers で PostgreSQL を使用
/// </summary>
type AutoJournalRepositoryTest() =
    inherit DatabaseTestBase()

    /// テストデータをクリーンアップ
    member private this.CleanupTestDataAsync() =
        task {
            use conn = new NpgsqlConnection(this.ConnectionString)
            do! conn.OpenAsync()

            use cmd = new NpgsqlCommand("""
                DELETE FROM "自動仕訳実行ログ";
                DELETE FROM "自動仕訳パターン明細";
                DELETE FROM "自動仕訳パターン";
            """, conn)
            let! _ = cmd.ExecuteNonQueryAsync()
            ()
        }

    /// テスト用パターンを作成
    member private _.CreateTestPattern(code: string, name: string, isActive: bool) : AutoJournalPattern =
        {
            Id = 0L
            PatternCode = code
            PatternName = name
            SourceTableName = "テストテーブル"
            Description = Some "テスト説明"
            IsActive = isActive
        }

    /// テスト用パターン明細を作成
    member private _.CreateTestPatternItem(patternId: int64, lineNo: int, dcType: DebitCreditType, accountCode: string) : AutoJournalPatternItem =
        {
            Id = 0L
            PatternId = patternId
            LineNumber = lineNo
            DebitCreditType = dcType
            AccountCode = AccountCode.Create(accountCode)
            AmountExpression = "{金額}"
            DescriptionTemplate = Some "テスト摘要"
        }

    [<Fact>]
    member this.``SavePatternAsync で新規パターンを保存できる``() =
        task {
            // Given
            do! this.CleanupTestDataAsync()
            use conn = new NpgsqlConnection(this.ConnectionString)
            do! conn.OpenAsync()

            let repository = AutoJournalRepository(conn) :> IAutoJournalRepository
            let pattern = this.CreateTestPattern("TEST001", "テストパターン", true)

            // When
            let! saved = repository.SavePatternAsync(pattern)

            // Then
            saved.Id |> should be (greaterThan 0L)
            saved.PatternCode |> should equal "TEST001"
            saved.PatternName |> should equal "テストパターン"
            saved.IsActive |> should equal true
        }

    [<Fact>]
    member this.``SavePatternAsync で既存パターンを更新できる``() =
        task {
            // Given
            do! this.CleanupTestDataAsync()
            use conn = new NpgsqlConnection(this.ConnectionString)
            do! conn.OpenAsync()

            let repository = AutoJournalRepository(conn) :> IAutoJournalRepository
            let pattern = this.CreateTestPattern("TEST001", "テストパターン", true)
            let! saved = repository.SavePatternAsync(pattern)

            // When
            let updated = { saved with PatternName = "更新後パターン"; IsActive = false }
            let! result = repository.SavePatternAsync(updated)

            // Then
            result.Id |> should equal saved.Id
            result.PatternName |> should equal "更新後パターン"
            result.IsActive |> should equal false
        }

    [<Fact>]
    member this.``GetPatternByCodeAsync で存在するパターンを取得できる``() =
        task {
            // Given
            do! this.CleanupTestDataAsync()
            use conn = new NpgsqlConnection(this.ConnectionString)
            do! conn.OpenAsync()

            let repository = AutoJournalRepository(conn) :> IAutoJournalRepository
            let pattern = this.CreateTestPattern("TEST001", "テストパターン", true)
            let! _ = repository.SavePatternAsync(pattern)

            // When
            let! result = repository.GetPatternByCodeAsync("TEST001")

            // Then
            result.IsSome |> should equal true
            result.Value.PatternCode |> should equal "TEST001"
            result.Value.PatternName |> should equal "テストパターン"
        }

    [<Fact>]
    member this.``GetPatternByCodeAsync で存在しないパターンは None を返す``() =
        task {
            // Given
            do! this.CleanupTestDataAsync()
            use conn = new NpgsqlConnection(this.ConnectionString)
            do! conn.OpenAsync()

            let repository = AutoJournalRepository(conn) :> IAutoJournalRepository

            // When
            let! result = repository.GetPatternByCodeAsync("NOTEXIST")

            // Then
            result.IsNone |> should equal true
        }

    [<Fact>]
    member this.``GetPatternsAsync で有効なパターンのみ取得できる``() =
        task {
            // Given
            do! this.CleanupTestDataAsync()
            use conn = new NpgsqlConnection(this.ConnectionString)
            do! conn.OpenAsync()

            let repository = AutoJournalRepository(conn) :> IAutoJournalRepository
            let! _ = repository.SavePatternAsync(this.CreateTestPattern("ACTIVE1", "有効1", true))
            let! _ = repository.SavePatternAsync(this.CreateTestPattern("INACTIVE1", "無効1", false))
            let! _ = repository.SavePatternAsync(this.CreateTestPattern("ACTIVE2", "有効2", true))

            // When
            let query: AutoJournalPatternQuery = { PatternCode = None; IsActive = Some true }
            let! patterns = repository.GetPatternsAsync(query)

            // Then
            patterns.Length |> should equal 2
            patterns |> List.forall (fun p -> p.IsActive) |> should equal true
        }

    [<Fact>]
    member this.``GetPatternsAsync で全パターンを取得できる``() =
        task {
            // Given
            do! this.CleanupTestDataAsync()
            use conn = new NpgsqlConnection(this.ConnectionString)
            do! conn.OpenAsync()

            let repository = AutoJournalRepository(conn) :> IAutoJournalRepository
            let! _ = repository.SavePatternAsync(this.CreateTestPattern("TEST1", "テスト1", true))
            let! _ = repository.SavePatternAsync(this.CreateTestPattern("TEST2", "テスト2", false))

            // When
            let query: AutoJournalPatternQuery = { PatternCode = None; IsActive = None }
            let! patterns = repository.GetPatternsAsync(query)

            // Then
            patterns.Length |> should equal 2
        }

    [<Fact>]
    member this.``SavePatternItemsAsync でパターン明細を保存できる``() =
        task {
            // Given
            do! this.CleanupTestDataAsync()
            use conn = new NpgsqlConnection(this.ConnectionString)
            do! conn.OpenAsync()

            let repository = AutoJournalRepository(conn) :> IAutoJournalRepository
            let! pattern = repository.SavePatternAsync(this.CreateTestPattern("TEST001", "テスト", true))

            let items = [
                this.CreateTestPatternItem(pattern.Id, 1, Debit, "130")
                this.CreateTestPatternItem(pattern.Id, 2, Credit, "500")
            ]

            // When
            do! repository.SavePatternItemsAsync pattern.Id items

            // Then
            let! savedItems = repository.GetPatternItemsAsync(pattern.Id)
            savedItems.Length |> should equal 2
            savedItems.[0].LineNumber |> should equal 1
            savedItems.[0].DebitCreditType |> should equal Debit
            savedItems.[1].LineNumber |> should equal 2
            savedItems.[1].DebitCreditType |> should equal Credit
        }

    [<Fact>]
    member this.``SavePatternItemsAsync で既存明細を置き換える``() =
        task {
            // Given
            do! this.CleanupTestDataAsync()
            use conn = new NpgsqlConnection(this.ConnectionString)
            do! conn.OpenAsync()

            let repository = AutoJournalRepository(conn) :> IAutoJournalRepository
            let! pattern = repository.SavePatternAsync(this.CreateTestPattern("TEST001", "テスト", true))

            // 初回の明細保存
            let items1 = [
                this.CreateTestPatternItem(pattern.Id, 1, Debit, "130")
                this.CreateTestPatternItem(pattern.Id, 2, Credit, "500")
            ]
            do! repository.SavePatternItemsAsync pattern.Id items1

            // When: 異なる明細で上書き
            let items2 = [
                this.CreateTestPatternItem(pattern.Id, 1, Debit, "111")
            ]
            do! repository.SavePatternItemsAsync pattern.Id items2

            // Then
            let! savedItems = repository.GetPatternItemsAsync(pattern.Id)
            savedItems.Length |> should equal 1
            savedItems.[0].AccountCode.Code |> should equal "111"
        }

    [<Fact>]
    member this.``GetPatternItemsAsync でパターン明細を取得できる``() =
        task {
            // Given
            do! this.CleanupTestDataAsync()
            use conn = new NpgsqlConnection(this.ConnectionString)
            do! conn.OpenAsync()

            let repository = AutoJournalRepository(conn) :> IAutoJournalRepository
            let! pattern = repository.SavePatternAsync(this.CreateTestPattern("TEST001", "テスト", true))

            let item1: AutoJournalPatternItem = {
                Id = 0L
                PatternId = pattern.Id
                LineNumber = 1
                DebitCreditType = Debit
                AccountCode = AccountCode.Create("130")
                AmountExpression = "{売上金額}"
                DescriptionTemplate = Some "売上計上"
            }
            let item2: AutoJournalPatternItem = {
                Id = 0L
                PatternId = pattern.Id
                LineNumber = 2
                DebitCreditType = Credit
                AccountCode = AccountCode.Create("500")
                AmountExpression = "{売上金額}"
                DescriptionTemplate = None
            }
            do! repository.SavePatternItemsAsync pattern.Id [item1; item2]

            // When
            let! savedItems = repository.GetPatternItemsAsync(pattern.Id)

            // Then
            savedItems.Length |> should equal 2
            savedItems.[0].AmountExpression |> should equal "{売上金額}"
            savedItems.[0].DescriptionTemplate |> should equal (Some "売上計上")
            savedItems.[1].DescriptionTemplate |> should equal None
        }

    [<Fact>]
    member this.``SaveLogAsync で実行ログを保存できる``() =
        task {
            // Given
            do! this.CleanupTestDataAsync()
            use conn = new NpgsqlConnection(this.ConnectionString)
            do! conn.OpenAsync()

            let repository = AutoJournalRepository(conn) :> IAutoJournalRepository
            let! pattern = repository.SavePatternAsync(this.CreateTestPattern("TEST001", "テスト", true))

            let log = AutoJournalLog.createSuccess pattern.Id 100 100

            // When
            let! saved = repository.SaveLogAsync(log)

            // Then
            saved.Id |> should be (greaterThan 0L)
            saved.PatternId |> should equal pattern.Id
            saved.ProcessedCount |> should equal 100
            saved.GeneratedCount |> should equal 100
            saved.Status |> should equal Success
        }

    [<Fact>]
    member this.``SaveLogAsync で失敗ログを保存できる``() =
        task {
            // Given
            do! this.CleanupTestDataAsync()
            use conn = new NpgsqlConnection(this.ConnectionString)
            do! conn.OpenAsync()

            let repository = AutoJournalRepository(conn) :> IAutoJournalRepository
            let! pattern = repository.SavePatternAsync(this.CreateTestPattern("TEST001", "テスト", true))

            let log = AutoJournalLog.createFailure pattern.Id 50 "処理エラー" "詳細なエラー内容"

            // When
            let! saved = repository.SaveLogAsync(log)

            // Then
            saved.Status |> should equal Failure
            saved.Message |> should equal (Some "処理エラー")
            saved.ErrorDetail |> should equal (Some "詳細なエラー内容")
        }

    [<Fact>]
    member this.``GetLogsByPatternIdAsync で実行ログを取得できる``() =
        task {
            // Given
            do! this.CleanupTestDataAsync()
            use conn = new NpgsqlConnection(this.ConnectionString)
            do! conn.OpenAsync()

            let repository = AutoJournalRepository(conn) :> IAutoJournalRepository
            let! pattern = repository.SavePatternAsync(this.CreateTestPattern("TEST001", "テスト", true))

            // 3件のログを保存
            let! _ = repository.SaveLogAsync(AutoJournalLog.createSuccess pattern.Id 10 10)
            let! _ = repository.SaveLogAsync(AutoJournalLog.createSuccess pattern.Id 20 20)
            let! _ = repository.SaveLogAsync(AutoJournalLog.createSuccess pattern.Id 30 30)

            // When
            let! logs = repository.GetLogsByPatternIdAsync pattern.Id 2

            // Then
            logs.Length |> should equal 2
        }

    [<Fact>]
    member this.``GetLogsByPatternIdAsync でログが実行日時の降順で返される``() =
        task {
            // Given
            do! this.CleanupTestDataAsync()
            use conn = new NpgsqlConnection(this.ConnectionString)
            do! conn.OpenAsync()

            let repository = AutoJournalRepository(conn) :> IAutoJournalRepository
            let! pattern = repository.SavePatternAsync(this.CreateTestPattern("TEST001", "テスト", true))

            // 異なる処理件数でログを保存（挿入順で時刻が異なる）
            let! _ = repository.SaveLogAsync(AutoJournalLog.createSuccess pattern.Id 10 10)
            let! _ = repository.SaveLogAsync(AutoJournalLog.createSuccess pattern.Id 20 20)
            let! _ = repository.SaveLogAsync(AutoJournalLog.createSuccess pattern.Id 30 30)

            // When
            let! logs = repository.GetLogsByPatternIdAsync pattern.Id 10

            // Then
            logs.Length |> should equal 3
            // 最新のログ（処理件数30）が最初に来る
            logs.[0].ProcessedCount |> should equal 30
        }
